c10/core/TensorImpl.h

```diff
@@ -129,6 +129,20 @@ struct C10_API PlacementDeleteContext {
   }
 };
 
+// One TensorImpl can be copied to another TensorImpl if they have the same
+// type_id. The only two special cases (for legacy reason) are:
+// CPUTensorId is compatible with CUDATensorId and SparseCPUTensorId is
+// compatible with SparseCUDATensorId.
+inline bool has_compatible_type_for_copy(
```

Comment:
can we move this somewhere else?  Either right next to shallow_copy_from (okay, it doesn't really belong there because it's not TensorImpl related, but you always call them together), or to where we usually do checking/translation, which I think is currently in `Backend.h` (those should probably be TensorTypeId based, but oh well...).

Reply:
moved into TensorImpl as inline method